### DISCUSSION TITLE
How to monitor Customer Editor Event?

### DISCUSSION THREAD

**Discussion by SunnyFitz** [eclipse-theia/theia#9733] [Q&A]
I want to monitor my customer editor event such as Open、Switch、Close…
I tried onwillexcuteCommand ,but click navigator file to open it donot active any command.
I tried widgetManage.onDidCreateWidget , it can get all widget create event without switch and close.
I tried editorManager.onActiveEditorChanged, it can get all Editor switch event without Create and close, but it can not monitor clicked in navigator tree when file already opened, and this cause editor switched which will not active editorManager.onActiveEditorChanged Event.
So if there any better way to monitor customer editor event like those interface in plugins （theia.workspace.ondidOpenTextDocumenmt 、ondidClose and ondidChange)?
Thanks

**Comment by vince-fugnitto** [comment]
@SunnyFitz thank you for the discussion, I see that you've already asked the question on [discourse](https://community.theia-ide.org/t/how-to-monitor-customer-editor-event/1800) (closed now), only one forum is necessary in the future :+1: 

> I want to monitor my customer editor event

I assume you meant a custom editor, is it an `extends` of `EditorWidget`?

Generally we use the [EditorManager](https://eclipse-theia.github.io/theia/docs/next/classes/editor.editormanager-1.html) to handle events of editors, which can tell us about when an editor is opened, closed, active editor is changed and so on.

For the `switch`, I understand you want to determine when something is clicked in the navigator/explorer even if it is already opened?

**Reply by SunnyFitz** [reply]
Yes, i asked same question in the theia community, but there's no replay.
i want to monitor event about editor include customer editor (maybe file event) ,i will distinguish them by myself.
i tried EditorManager but only get active editor change event. How to get opened closed event ?

**Reply by vince-fugnitto** [reply]
> Yes, i asked same question in the theia community, but there's no replay.

Someone will try to reply when they get a chance or know an answer, I believe it was open only for a day.

> How to get opened closed event ?

You can use the `EditorManager` to know when an editor is created, disposed (removed), and activated. Would that not be enough for your use-case? For example:

https://github.com/eclipse-theia/theia/blob/8dd8401c8125ab679de727f6576d76cea4d89d2c/packages/editor/src/browser/editor-navigation-contribution.ts#L63-L71

Here we listen to when an editor is opened, and we are notified when it is disposed (removed) so we can handle it.

**Reply by SunnyFitz** [reply]
Thank you for reply . The editorManager.onCurrentEditorChanged can get event of open and switch.  How to get close event about editor ？And i found editormanager can only get code editor event. It cant get customEditor Event.

**Reply by vince-fugnitto** [reply]
For `close` you can take a look at the snippet I provided, when the editor is `disposed` it means that it has been removed/closed.

I asked this question previously without a reply but can you clarify what the `custom editor` is you are referring to?
- is it from vscode extensions?
- is it an `extends EditorWidget`?
